nerium.py

```diff
@@ -1,97 +1,142 @@
-    in $DATABASE_URL. Returns results as dict formatted like
+    @abc.abstractmethod
+    def result(self):
+        """ Return query results as a JSON-serializable Python structure
+        (generally a list of dictionaries)
+        """
+        return
 
-    `{"columns": [<list of column names>], "data": [<list of rows as lists>]}`
 
-    Args:
-        report_name (str): Name of chosen query file, without the .sql extension
+class SQLResultSet(ResultSet):
+    """ SQL database query implementation of ResultSet, using Records library
+    to fetch a dataset from configured report_name
     """
-    # Find SQL file matching report_name
-    def get_sql(self, report_name):
-        try:
-            sql_dir = current_app.config['SQL_PATH']
-        except RuntimeError:  # so this works outside requests for testing
-            sql_dir = os.getenv('SQL_PATH', 'sqls')
-        filename = '{}.sql'.format(report_name)
-        sql_file = os.path.join(sql_dir, filename)
-        return sql_file
-
-    # Submit query, apply serialization cleaner method to data values,
-    #    return results in dict
-    def result(self, report_name, **kwargs):
+
+    @property
+    def query_type(self):
+        return ('sql')
+
+    query_db = records.Database(app.config['DATABASE_URL'])
+
+    def result(self):
         try:
-            query_db = records.Database(current_app.config['DATABASE_URL'])
-        except RuntimeError:
-            query_db = records.Database(os.getenv('DATABASE_URL', 'sqlite:///'))
-        result = query_db.query_file(self.get_sql(report_name), **kwargs)
-        cols = result.dataset.headers
-        rows = [list(row.values()) for row in result.all(as_ordereddict=True)]
-        format_dict = dict(columns=cols, data=rows)
-        return format_dict
+            rows = self.query_db.query_file(self.get_file(), **self.kwargs)
+            result = rows.as_dict()
+        except IOError as e:
+            result = [{'Error': e, }, ]
+        return result
 
 
-class ReportAPI(Resource):
-    """ Calls ResultSet.result() and returns JSON in compact format like
+class ResultFormatter(ABC):
+    def __init__(self, result):
+        self.result = result
 
-    `{"columns": [<list of column names>],
-      "data": [<array of row value arrays>]}`
+    @abc.abstractmethod
+    def format_results(self):
+        """ Transform result set structure and return new structure
+        """
+        return
+
+
+class CompactFormatter(ResultFormatter):
+    """ Returns dict in format
+    {"columns": [<list of column names>], "data": [<array of row value arrays>]}
     """
-    def get(self, report_name):
-        loader = ResultSet()
-        payload = loader.result(report_name, **request.args.to_dict())
-        return payload
+    # TODO: some graceful error handling, in case of bad input format, etc.
+    def format_results(self):
+        raw = self.result
+        columns = list(raw[0].keys())
```

Comment:
This will return keys in a random order, so I don't think the compact view will necessarily return the correct column matching.